astroquery/mast/observations.py

```diff
@@ -1129,53 +1186,57 @@ def mast_query(self, service, columns=None, **kwargs):
-        position_keys = ['ra', 'dec', 'radius', 'position']
-        request_keys = ['format', 'data', 'filename', 'timeout', 'clearcache',
-                        'removecache', 'removenullcolumns', 'page', 'pagesize']
+        position_keys = {'ra', 'dec', 'radius', 'position'}
+        request_keys = {'format', 'data', 'filename', 'timeout', 'clearcache',
+                        'removecache', 'removenullcolumns', 'page', 'pagesize'}
 
-        # Explicit formatting for Mast's filtered services
-        if 'filtered' in service.lower():
+        # Split params into categories
+        position_params = {k: v for k, v in kwargs.items() if k.lower() in position_keys}
+        request_params = {k: v for k, v in kwargs.items() if k.lower() in request_keys}
+        service_params = {k: v for k, v in kwargs.items() if k.lower() not in position_keys | request_keys}
 
-            # Separating the filter params from the positional and service_request method params.
-            filters = [{'paramName': k, 'values': kwargs[k]} for k in kwargs
-                       if k.lower() not in position_keys+request_keys]
-            position_params = {k: v for k, v in kwargs.items() if k.lower() in position_keys}
-            request_params = {k: v for k, v in kwargs.items() if k.lower() in request_keys}
+        # Handle filtered vs. non-filtered services
+        if 'filtered' in service.lower():
+            filters = self._build_filters(service_params)
 
-            # Mast's filtered services require at least one filter
-            if filters == []:
-                raise InvalidQueryError("Please provide at least one filter.")
+            if not filters:
+                raise InvalidQueryError('Please provide at least one filter.')
 
-            # Building 'params' for Mast.service_request
-            if columns is None:
-                columns = '*'
+            if columns is not None and isinstance(columns, list):
+                columns = ','.join(columns)
 
-            params = {'columns': columns,
-                      'filters': filters,
-                      **position_params
-                      }
+            params = {
+                'columns': columns or '*',
+                'filters': filters,
+                **position_params,
+            }
         else:
-
-            # Separating service specific params from service_request method params
-            params = {k: v for k, v in kwargs.items() if k.lower() not in request_keys}
-            request_params = {k: v for k, v in kwargs.items() if k.lower() in request_keys}
-
-            # Warning for wrong input
             if columns is not None:
-                warnings.warn("'columns' parameter will not mask non-filtered services", InputWarning)
+                warnings.warn(
+                    "'columns' parameter is ignored for non-filtered services.",
+                    InputWarning
+                )
```

Comment:
no need to change it back but in general we allow longer lines (120), so no need to cut up perfectly valid one liners